### PR TITLE
Uploader: init + close slice

### PR DESCRIPTION
**Pull request by michaljurecko**
Part of: https://keboola.atlassian.net/browse/BA-40

Changes:
- Implemented first part of the uploader: slice close operation

https://github.com/keboola/keboola-as-code/blob/fc3bb065207a369e96bd82cbddaf455f2bc848ee/internal/pkg/service/buffer/worker/upload/close.go#L17-L22

### CODE REVIEW

File: internal/pkg/service/buffer/worker/distribution/listener.go [hunk 1]
```diff
@@ -139,13 +142,21 @@ func (v *listeners) trigger() {
...
+}
+
 func (l *Listener) Stop() {
 	l.all.lock.Lock()
 	defer l.all.lock.Unlock()
 
 	l.cancel()
 	l.wg.Wait()
-	close(l.C)
 	delete(l.all.listeners, l.id)
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/buffer/worker/distribution/listener.go:160]: O channel sa postara garbage collector, ak sa uz nebude pouzivat:
https://stackoverflow.com/questions/8593645/is-it-ok-to-leave-a-channel-open

File: internal/pkg/service/buffer/worker/upload/close.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+				slice := event.Value
+				if _, err := u.store.SetSliceState(ctx, &slice, slicestate.Uploading); err != nil {
+					return "", err
+				}
+
+				return "slice closed", nil
+			}
+		},
+	})
+}
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/worker/upload/close.go:46]: A toto je klucova cast PR, prva cast `uploader`, kod co caka na zatvorenie danej slice (aby ziadna API node do slice uz nezapisovala), a prepne ju zo stavu `closing` -> `uploading`.

Samotny upload davam do dalsieho PR.

File: internal/pkg/service/buffer/worker/upload/close_test.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+[task][%s]DEBUG  lock acquired "runtime/lock/task/00000123/my-receiver/my-export/slice.close/0001-01-01T00:00:01.000Z/0001-01-01T00:00:01.000Z"
+[task][%s]INFO  waiting until all API nodes switch to a revision >= %s
+INFO  ---> first unlock
+INFO  ---> second unlock
+[watcher][worker]INFO  revision updated to "%s", unblocked "1" listeners
+[task][%s]INFO  task succeeded (30s): slice closed
+[task][%s]DEBUG  lock released "runtime/lock/task/00000123/my-receiver/my-export/slice.close/0001-01-01T00:00:01.000Z/0001-01-01T00:00:01.000Z"
+%A
+INFO  exited
+`, workerDeps.DebugLogger().AllMessages())
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/worker/upload/close_test.go:213]: Tu je v teste pekne vidiet co sa deje v API a Worker node pocas cakania na zatvorenie slice.

File: internal/pkg/service/buffer/watcher/apinode/apinode.go [hunk 4]
```diff
@@ -14,14 +16,24 @@ import (
...
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/prefixtree"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/common/servicectx"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
 )
 
-// APINode contains cached state of all configuration objects required by the import endpoint.
-type APINode struct {
+// Node contains cached state of all configuration objects required by the import endpoint, see GetReceiver method.
+type Node struct {
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/buffer/watcher/apinode/apinode.go:27]: Prepac, ze tie prve commits su menej prehladne.

Menil som, co som potreboval na dokoncenie `uploader`, alebo na to, aby sa to dalo otestovat.

**Review by JakubMatejka — APPROVED** [review]

**Pull request merged**